lib/rubygems.rb

```diff
@@ -119,6 +119,10 @@ module Gem
   # to avoid deprecation warnings in Ruby 2.7.
   UNTAINT = RUBY_VERSION < '2.7' ? :untaint.to_sym : proc{}
 
+  # When https://bugs.ruby-lang.org/issues/17259 is available, there is no need to override Kernel#warn
+  KERNEL_WARN_IGNORES_INTERNAL_ENTRIES = RUBY_ENGINE == "truffleruby" ||
```

Comment:
Isn't there any different way (without hardcoding engine and version) to check this functionality is present? If not, can't we do some sanity check (not printed to stdout) to check how warn behaves to detect?

Reply:
We would have to temporarily switch `$stderr` to test that (and to what? we don't want to load StringIO just for that).
RubyGems could be required potentially concurrently in some thread, so that change could end up breaking something else.
I think it's difficult to feature-detect this one.

If we want something more opt-in/flexible we could let `require "rubygems/defaults/#{RUBY_ENGINE}"` opt in for this behavior.
Then we'd need e.g. some `@ivar` on Gem and a method to set it like `Gem.warn_ignores_internal_entries`.

Reply:
I thought the same thing but this thing felt hard to feature-detect.

Reply:
OK as-is, or do you want I make it opt-in from `require "rubygems/defaults/#{RUBY_ENGINE}"`?
If the latter, could you decide the method names? (to set and to query)

Reply:
As a small note, I wrote https://github.com/rubygems/rubygems/pull/4075/files#r536002463 before reading https://github.com/rubygems/rubygems/pull/4075/files#r536001783. It was a reply to @simi's comment, not to @eregon's explanation. I'm :+1: with the hardcoded version/implementation check.

Reply:
Let's leave it as it then since there was no further objections.